upstairs/src/volume.rs

```diff
@@ -0,0 +1,1228 @@
+            sub_volume.conditional_activate(gen)?;
+
+            let sub_volume_computed_size = self.block_size
+                * (sub_volume.lba_range.end - sub_volume.lba_range.start);
+
+            if sub_volume.total_size()? != sub_volume_computed_size {
+                crucible_bail!(SubvolumeSizeMismatch);
+            }
+        }
+
+        if let Some(read_only_parent) = &self.read_only_parent {
+            read_only_parent.conditional_activate(gen)?;
+        }
+
+        Ok(())
+    }
+
+    fn query_is_active(&self) -> Result<bool, CrucibleError> {
+        for sub_volume in &self.sub_volumes {
+            if !sub_volume.query_is_active()? {
+                return Ok(false);
+            }
+        }
+
+        if let Some(read_only_parent) = &self.read_only_parent {
+            if !read_only_parent.query_is_active()? {
+                return Ok(false);
+            }
+        }
+
+        Ok(true)
+    }
+
+    fn total_size(&self) -> Result<u64, CrucibleError> {
+        if !self.sub_volumes.is_empty() {
+            // If this volume has sub volumes, compute total size based on those
+            let mut total_blocks = 0;
+
+            for sub_volume in &self.sub_volumes {
+                // Range is [start, end), meaning 0..10 is 10
+                total_blocks += (sub_volume.lba_range.end
+                    - sub_volume.lba_range.start)
+                    as u64;
+            }
+
+            Ok(total_blocks * self.block_size)
+        } else if let Some(read_only_parent) = &self.read_only_parent {
+            // If this volume only has a read only parent, report that size for
+            // total size
+            let total_blocks = read_only_parent.lba_range.end
+                - read_only_parent.lba_range.start;
+            Ok(total_blocks * self.block_size)
+        } else {
+            // If this volume has neither, then total size is 0
+            Ok(0)
+        }
+    }
+
+    fn get_block_size(&self) -> Result<u64, CrucibleError> {
+        Ok(self.block_size)
+    }
+
+    fn get_uuid(&self) -> Result<Uuid, CrucibleError> {
+        Ok(self.uuid)
+    }
+
+    fn read(
+        &self,
+        offset: Block,
+        data: Buffer,
+    ) -> Result<BlockReqWaiter, CrucibleError> {
+        // In the cast that this volume only has a read only parent, serve
+        // reads directly from that
+        if self.sub_volumes.is_empty() {
+            if let Some(read_only_parent) = &self.read_only_parent {
+                return read_only_parent.read(offset, data.clone());
+            } else {
+                crucible_bail!(
+                    CannotServeBlocks,
+                    "No read only parent, no sub volumes!",
+                );
+            }
+        }
+
+        let affected_sub_volumes = self.sub_volumes_for_lba_range(
+            offset.value,
+            data.len() as u64 / self.block_size,
+        );
+
+        // TODO parallel dispatch!
```

Comment:
Let's go make issues for all these TODOs.  We can move forward with this code and be sure to have issues so we don't forget the remaining pieces.